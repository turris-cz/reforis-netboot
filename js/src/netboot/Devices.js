/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useState } from "react";
import PropTypes from "prop-types";
import { API_STATE, Spinner, ErrorMessage } from "foris";

import DevicesTable from "./DevicesTable";
import {
    useLoadDevices,
    useAcceptDevice,
    useUpdateOnAccept,
    useUnpairDevice,
    useUpdateOnUnpair,
} from "./hooks";

Devices.propTypes = {
    ws: PropTypes.object.isRequired,
};

export default function Devices({ ws }) {
    const [devices, setDevices] = useState([]);

    const loadDevicesState = useLoadDevices(setDevices);

    const [acceptState, acceptDevice] = useAcceptDevice();
    useUpdateOnAccept(ws, setDevices);

    const [unpairState, unpairDevice] = useUnpairDevice();
    useUpdateOnUnpair(ws, setDevices);

    if (
        loadDevicesState === API_STATE.INIT ||
        [loadDevicesState, acceptState, unpairState].includes(API_STATE.SENDING)
    ) {
        return <Spinner />;
    }
    if (loadDevicesState === API_STATE.ERROR) {
        return <ErrorMessage />;
    }
    return (
        <DevicesTable
            devices={devices}
            acceptDevice={acceptDevice}
            unpairDevice={unpairDevice}
        />
    );
}
