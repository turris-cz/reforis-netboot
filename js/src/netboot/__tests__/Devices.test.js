/*
 * Copyright (C) 2020-2024 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import mockAxios from "jest-mock-axios";
import {
    render,
    act,
    wait,
    waitForElement,
    getByText,
    getByTitle,
    getAllByTitle,
    getByRole,
    queryByText,
    fireEvent,
} from "foris/testUtils/customTestRender";
import { mockJSONError } from "foris/testUtils/network";
import { mockSetAlert } from "foris/testUtils/alertContextMock";
import { WebSockets } from "foris";

import devices from "./__fixtures__/devices";
import Devices from "../Devices";

describe("<Devices />", () => {
    let container;
    let webSockets;

    beforeEach(() => {
        webSockets = new WebSockets();
        ({ container } = render(<Devices ws={webSockets} />));
    });

    it("should render spinner", () => {
        expect(container).toMatchSnapshot();
    });

    it("should render table", async () => {
        expect(mockAxios.get).toBeCalledWith(
            "/reforis/netboot/api/devices",
            expect.anything()
        );
        mockAxios.mockResponse({ data: devices });
        await waitForElement(() => getByText(container, devices[0].serial));
        expect(container).toMatchSnapshot();
    });

    it("should handle GET error", async () => {
        mockJSONError();
        await wait(() =>
            expect(
                getByText(container, "An error occurred while fetching data.")
            ).toBeTruthy()
        );
    });

    describe("accept", () => {
        let acceptButton;

        beforeEach(async () => {
            mockAxios.mockResponse({ data: devices });
            await waitForElement(() => getByText(container, devices[0].serial));
            acceptButton = getByText(container, "Accept pairing request");
        });

        it("should handle API error", async () => {
            // Unpair device
            fireEvent.click(acceptButton);
            expect(mockAxios.put).toBeCalledWith(
                `/reforis/netboot/api/accept/${devices[0].serial}`,
                undefined,
                expect.anything()
            );
            // Handle error
            const errorMessage = "API didn't handle this well";
            mockJSONError(errorMessage);
            await wait(() => {
                expect(mockSetAlert).toHaveBeenCalledWith(errorMessage);
            });
            // expect icon <i> with class .fa to be defined
            expect(document.querySelector(".fa")).toBeDefined();
        });

        it("should display spinner during request", async () => {
            fireEvent.click(acceptButton);
            expect(container).toMatchSnapshot();
        });

        it("should display spinner while processing acceptance request", async () => {
            fireEvent.click(acceptButton);
            mockAxios.mockResponse({ data: { task_id: "5542" } });
            await waitForElement(() => getByText(container, devices[0].serial));
            act(() =>
                webSockets.dispatch({
                    module: "netboot",
                    action: "accept",
                    data: { serial: devices[0].serial, status: "started" },
                })
            );
            expect(getByRole(container, "status")).toBeDefined();
        });

        it("should handle succesfully processed acceptance request", async () => {
            expect(document.querySelector(".fa")).toBeDefined();
            fireEvent.click(acceptButton);
            mockAxios.mockResponse({ data: { task_id: "5542" } });
            await waitForElement(() => getByText(container, devices[0].serial));
            act(() =>
                webSockets.dispatch({
                    module: "netboot",
                    action: "accept",
                    data: { serial: devices[0].serial, status: "succeeded" },
                })
            );
            expect(document.querySelector(".fa")).toBeDefined();
        });

        it("should handle failed acceptance request", async () => {
            fireEvent.click(acceptButton);
            mockAxios.mockResponse({ data: { task_id: "5542" } });
            await waitForElement(() => getByText(container, devices[0].serial));
            act(() =>
                webSockets.dispatch({
                    module: "netboot",
                    action: "accept",
                    data: { serial: devices[0].serial, status: "failed" },
                })
            );
            expect(mockSetAlert).toHaveBeenCalledWith("Cannot pair devices.");
            // Request didn't change its status
            expect(document.querySelector(".fa")).toBeDefined();
        });
    });

    describe("unpair", () => {
        let unpairButton;

        beforeEach(async () => {
            mockAxios.mockResponse({ data: devices });
            await waitForElement(() => getByText(container, devices[1].serial));
            unpairButton = getByText(container, "Unpair device");
        });

        it("should handle API error", async () => {
            // Unpair device
            fireEvent.click(unpairButton);
            expect(mockAxios.put).toBeCalledWith(
                `/reforis/netboot/api/unpair/${devices[1].serial}`,
                undefined,
                expect.anything()
            );
            // Handle error
            const errorMessage = "API didn't handle this well";
            mockJSONError(errorMessage);
            await wait(() => {
                expect(mockSetAlert).toHaveBeenCalledWith(errorMessage);
            });
        });

        it("should display spinner during request", async () => {
            fireEvent.click(unpairButton);
            expect(container).toMatchSnapshot();
        });

        it("should remove device from table on success", async () => {
            fireEvent.click(unpairButton);
            mockAxios.mockResponse({ data: { result: true } });
            await waitForElement(() => getByText(container, devices[1].serial));
            act(() =>
                webSockets.dispatch({
                    module: "netboot",
                    action: "revoke",
                    data: { serial: devices[1].serial },
                })
            );
            await wait(() =>
                expect(queryByText(container, devices[1].serial)).toBeNull()
            );
        });
    });
});
