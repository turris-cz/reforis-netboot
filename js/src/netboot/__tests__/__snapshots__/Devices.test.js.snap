// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Devices /> accept should display spinner during request 1`] = `
<div>
  <div
    class="spinner-wrapper my-3 text-center"
  >
    <div
      class="spinner-border"
      role="status"
    >
      <span
        class="sr-only"
      />
    </div>
  </div>
</div>
`;

exports[`<Devices /> should render spinner 1`] = `
<div>
  <div
    class="spinner-wrapper my-3 text-center"
  >
    <div
      class="spinner-border"
      role="status"
    >
      <span
        class="sr-only"
      />
    </div>
  </div>
</div>
`;

exports[`<Devices /> should render table 1`] = `
<div>
  <div
    class="table-responsive"
  >
    <table
      class="table table-hover netboot-devices-table"
    >
      <thead>
        <tr>
          <th
            class="netboot-serial-number"
            scope="col"
          >
            Serial Number
          </th>
          <th
            class="netboot-status"
            scope="col"
          >
            Paired
          </th>
          <th
            aria-label="Change status"
            class="netboot-action"
            scope="col"
          />
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            1234
          </td>
          <td
            class="text-center"
          >
            <span>
              <i
                class="fa fa-lg fa-times-circle text-primary"
                title="Awaiting acceptance"
              />
            </span>
          </td>
          <td
            class="text-center"
          >
            <button
              class="btn btn-primary "
              type="button"
            >
               
               
              Accept pairing request
            </button>
          </td>
        </tr>
        <tr>
          <td>
            5678
          </td>
          <td
            class="text-center"
          >
            <span>
              <i
                class="fa fa-lg fa-check-circle text-success"
                title="Paired"
              />
            </span>
          </td>
          <td
            class="text-center"
          >
            <button
              class="btn btn-primary "
              type="button"
            >
               
               
              Unpair device
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`<Devices /> unpair should display spinner during request 1`] = `
<div>
  <div
    class="spinner-wrapper my-3 text-center"
  >
    <div
      class="spinner-border"
      role="status"
    >
      <span
        class="sr-only"
      />
    </div>
  </div>
</div>
`;
